os/20201103-full-disk-raid.md

```diff
@@ -0,0 +1,208 @@
+      }
+    ],
+    "filesystems": [
+      {
+        "device": "/dev/disk/by-partlabel/esp-1",
+        "format": "vfat",
+        "label": "esp-1",
+        "wipeFilesystem": true
+      },
+      {
+        "device": "/dev/disk/by-partlabel/esp-2",
+        "format": "vfat",
+        "label": "esp-2",
+        "wipeFilesystem": true
+      },
+      {
+        "device": "/dev/md/md-boot",
+        "format": "ext4",
+        "label": "boot",
+        "wipeFilesystem": true
+      },
+      {
+        "device": "/dev/md/md-root",
+        "format": "xfs",
+        "label": "root",
+        "wipeFilesystem": true
+      }
+    ],
+    "raid": [
+      {
+        "devices": [
+          "/dev/disk/by-partlabel/boot-1",
+          "/dev/disk/by-partlabel/boot-2"
+        ],
+        "level": "raid1",
+        "name": "md-boot",
+        "options": [
+          "--metadata=1.0"
+        ]
+      },
+      {
+        "devices": [
+          "/dev/disk/by-partlabel/root-1",
+          "/dev/disk/by-partlabel/root-2"
+        ],
+        "level": "raid1",
+        "name": "md-root"
+      }
+    ]
+  }
+}
+```
+
+FCCT assigns a unique serial number to each disk and appends it to the label assigned to each disk partition.  This ensures each partition can be referenced by a unique path in `/dev/disk/by-label`.
+
+FCCT also hardcodes the partition size for each partition, replicating the existing partition layout.  If we change the default partition sizes in future, we can encode them in a new layout, and we can also change the default layout if we bump the FCC major version.
+
+The original boot disk is not treated specially; it's wiped and rewritten just like any other replica.
+
+## Root filesystem
+
+The Dracut glue already saves and restores the root filesystem contents whenever the config specifies a filesystem with label `root` and `wipeFilesystem` set to `true`.
+
+## Boot filesystem
+
+Similar to the root filesystem, the Dracut glue will save and restore the contents of `/boot` whenever the config specifies a filesystem labeled `boot` with `wipeFilesystem` set to `true`.
+
+GRUB includes a module for reading MD-RAID volumes.  However, the shipped boot sector is hardcoded to set `prefix` to the first boot disk, and we want to replicate the boot sector via a bit-for-bit copy (see below).  Therefore, we'll create the RAID volume with metadata format 1.0, which puts the RAID superblock at the end of the partition.  This allows GRUB to treat individual `/boot` replicas as independent filesystems, provided that the RAID module is _not_ preloaded into the BIOS boot partition.  (It currently is not.)
+
+`/boot` replication prevents the use of the grubenv mechanism.  If GRUB treats the replicas as independent filesystems, grubenv will cause the RAID to desynchronize.  Alternatively, if we were to use the GRUB RAID module, GRUB would [disable grubenv support](https://www.gnu.org/software/grub/manual/grub/grub.html#Environment-block).
```

Comment:
It'd be helpful to see what RHEL/Fedora does here. Have you had a chance to look there?

Reply:
Neither CentOS 8 nor Fedora 33 provide templates in Anaconda for setting up RAID.  When doing it by hand in the Anaconda partitioner, it's possible to set up RAIDed `/boot` and `/boot/efi`.  In that case, Anaconda automatically uses a RAID 1.0 superblock for the ESP but not for `/boot`.

Also of note, Anaconda doesn't provide a mechanism for creating partitions inside a RAID device.